geniza/corpus/models.py

```diff
@@ -956,6 +956,19 @@ def digital_translations(self):
             doc_relation__contains=Footnote.DIGITAL_TRANSLATION
         ).order_by("source")
 
+    @property
+    def default_translation(self):
+        """The first translation footnote that is in the current language, or the first
+        translation footnote ordered alphabetically by source if one is not available
+        in the current language."""
+
+        translations = self.digital_translations()
+        in_language = translations.filter(source__languages__code=get_language())
```

Comment:
IDK if it matters here but I think filtering incurs an additional db query

Reply:
Curiously, it seems to be five queries either way, just with an `INNER JOIN` when filtering. Without that line, and just picking `translations.first()`:

```sql
[20/Jul/2023 14:30:32] DEBUG:django.db.backends::(0.001) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."app_label" = 'corpus' AND "django_content_type"."model" = 'document') LIMIT 21; args=('corpus', 'document')
[20/Jul/2023 14:30:32] DEBUG:django.db.backends::(0.006) SELECT "footnotes_footnote"."id", "footnotes_footnote"."source_id", "footnotes_footnote"."location", "footnotes_footnote"."location_sort", "footnotes_footnote"."doc_relation", "footnotes_footnote"."notes", "footnotes_footnote"."content", "footnotes_footnote"."url", "footnotes_footnote"."content_type_id", "footnotes_footnote"."object_id" FROM "footnotes_footnote" INNER JOIN "footnotes_source" ON ("footnotes_footnote"."source_id" = "footnotes_source"."id") WHERE ("footnotes_footnote"."content_type_id" = 8 AND "footnotes_footnote"."object_id" = 17210 AND "footnotes_footnote"."doc_relation"::text LIKE '%Y%') ORDER BY "footnotes_source"."title" ASC, "footnotes_source"."year" ASC LIMIT 1; args=(8, 17210, '%Y%')
[20/Jul/2023 14:30:32] DEBUG:django.db.backends::(0.000) SELECT "corpus_document"."doc_date_original", "corpus_document"."doc_date_calendar", "corpus_document"."doc_date_standard", "corpus_document"."id", "corpus_document"."image_order_override", "corpus_document"."shelfmark_override", "corpus_document"."description", "corpus_document"."description_en", "corpus_document"."description_he", "corpus_document"."description_ar", "corpus_document"."doctype_id", "corpus_document"."language_note", "corpus_document"."notes", "corpus_document"."created", "corpus_document"."last_modified", "corpus_document"."needs_review", "corpus_document"."old_pgpids", "corpus_document"."status" FROM "corpus_document" WHERE "corpus_document"."id" = 17210 LIMIT 21; args=(17210,)
[20/Jul/2023 14:30:32] DEBUG:django.db.backends::(0.001) SELECT "corpus_textblock"."id", "corpus_textblock"."document_id", "corpus_textblock"."fragment_id", "corpus_textblock"."certain", "corpus_textblock"."selected_images", "corpus_textblock"."region", "corpus_textblock"."multifragment", "corpus_textblock"."order" FROM "corpus_textblock" WHERE "corpus_textblock"."document_id" = 17210 ORDER BY "corpus_textblock"."order" ASC; args=(17210,)
[20/Jul/2023 14:30:32] DEBUG:django.db.backends::(0.001) SELECT "corpus_fragment"."id", "corpus_fragment"."shelfmark", "corpus_fragment"."old_shelfmarks", "corpus_fragment"."collection_id", "corpus_fragment"."url", "corpus_fragment"."iiif_url", "corpus_fragment"."is_multifragment", "corpus_fragment"."notes", "corpus_fragment"."needs_review", "corpus_fragment"."manifest_id", "corpus_fragment"."created", "corpus_fragment"."last_modified" FROM "corpus_fragment" WHERE "corpus_fragment"."id" = 17929 LIMIT 21; args=(17929,)
```

as written:

```sql
[20/Jul/2023 14:31:21] DEBUG:django.db.backends::(0.001) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."app_label" = 'corpus' AND "django_content_type"."model" = 'document') LIMIT 21; args=('corpus', 'document')
[20/Jul/2023 14:31:21] DEBUG:django.db.backends::(0.007) SELECT "footnotes_footnote"."id", "footnotes_footnote"."source_id", "footnotes_footnote"."location", "footnotes_footnote"."location_sort", "footnotes_footnote"."doc_relation", "footnotes_footnote"."notes", "footnotes_footnote"."content", "footnotes_footnote"."url", "footnotes_footnote"."content_type_id", "footnotes_footnote"."object_id" FROM "footnotes_footnote" INNER JOIN "footnotes_source" ON ("footnotes_footnote"."source_id" = "footnotes_source"."id") INNER JOIN "footnotes_source_languages" ON ("footnotes_source"."id" = "footnotes_source_languages"."source_id") INNER JOIN "footnotes_sourcelanguage" ON ("footnotes_source_languages"."sourcelanguage_id" = "footnotes_sourcelanguage"."id") WHERE ("footnotes_footnote"."content_type_id" = 8 AND "footnotes_footnote"."object_id" = 17210 AND "footnotes_footnote"."doc_relation"::text LIKE '%Y%' AND "footnotes_sourcelanguage"."code" = 'en') ORDER BY "footnotes_source"."title" ASC, "footnotes_source"."year" ASC LIMIT 1; args=(8, 17210, '%Y%', 'en')
[20/Jul/2023 14:31:21] DEBUG:django.db.backends::(0.000) SELECT "corpus_document"."doc_date_original", "corpus_document"."doc_date_calendar", "corpus_document"."doc_date_standard", "corpus_document"."id", "corpus_document"."image_order_override", "corpus_document"."shelfmark_override", "corpus_document"."description", "corpus_document"."description_en", "corpus_document"."description_he", "corpus_document"."description_ar", "corpus_document"."doctype_id", "corpus_document"."language_note", "corpus_document"."notes", "corpus_document"."created", "corpus_document"."last_modified", "corpus_document"."needs_review", "corpus_document"."old_pgpids", "corpus_document"."status" FROM "corpus_document" WHERE "corpus_document"."id" = 17210 LIMIT 21; args=(17210,)
[20/Jul/2023 14:31:21] DEBUG:django.db.backends::(0.001) SELECT "corpus_textblock"."id", "corpus_textblock"."document_id", "corpus_textblock"."fragment_id", "corpus_textblock"."certain", "corpus_textblock"."selected_images", "corpus_textblock"."region", "corpus_textblock"."multifragment", "corpus_textblock"."order" FROM "corpus_textblock" WHERE "corpus_textblock"."document_id" = 17210 ORDER BY "corpus_textblock"."order" ASC; args=(17210,)
[20/Jul/2023 14:31:21] DEBUG:django.db.backends::(0.001) SELECT "corpus_fragment"."id", "corpus_fragment"."shelfmark", "corpus_fragment"."old_shelfmarks", "corpus_fragment"."collection_id", "corpus_fragment"."url", "corpus_fragment"."iiif_url", "corpus_fragment"."is_multifragment", "corpus_fragment"."notes", "corpus_fragment"."needs_review", "corpus_fragment"."manifest_id", "corpus_fragment"."created", "corpus_fragment"."last_modified" FROM "corpus_fragment" WHERE "corpus_fragment"."id" = 17929 LIMIT 21; args=(17929,)
```

Reply:
oh, of course, it requires the extra queries to get the data no matter what because we haven't prefetched anything. 

Five queries seems like a lot for this, might be worth adding some prefetching; probably reasonable to add to the `digital_translations` method (or somewhere shared by digital translations and digital editions methods?), since we're probably always going to want source and language information when we pull the translations.

Reply:
I can't seem to reduce the number of queries with prefetching, but maybe I'm doing it wrong? (Definitely not my strong suit!)

```py
    def digital_translations(self):
        """All footnotes for this document where the document relation includes
        digital translation."""

        return (
            self.footnotes.select_related("source")
            .prefetch_related(
                "source__authorship_set__creator",
                "source__languages",
            )
            .filter(doc_relation__contains=Footnote.DIGITAL_TRANSLATION)
            .order_by("source")
        )
```

This now results in 8 queries for `digital_translations()` and 7 queries for `default_translation` afterwards (8 if run without `digital_translations()` first).

Reply:
ok, did a little experimenting in the console - if you're only filtering on source language, then that's the only thing you'd need to fetch. Something like this: 

`doc.footnotes.all().select_related('source').prefetch_related('source__languages')`

In the console, it looks like two queries to me - one to get the footnotes + select related, and one for the prefetch, which always requires a separate query.

There are some `metadata_prefetch` methods on the `Source` and `Footnote` querysets that we used for the data exports, but it seems like it's more than you need here (unless you do need creator? in which case maybe you can just use that).   I'm not sure if there's a good way to share the footnotes with prefetching between translation and edition methods, or if django does that for you automatically.

Reply:
Curious, I'm still getting more queries by prefetching like that—the one additional prefetch query in the original method, but then also an additional one in `default_translation` after having called it once.

without prefetching:

```sql
>>> d.digital_translations()
[08/Aug/2023 15:19:35] DEBUG:django.db.backends::(0.001) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."app_label" = 'corpus' AND "django_content_type"."model" = 'document') LIMIT 21; args=('corpus', 'document')
[08/Aug/2023 15:19:35] DEBUG:django.db.backends::(0.006) SELECT "footnotes_footnote"."id", "footnotes_footnote"."source_id", "footnotes_footnote"."location", "footnotes_footnote"."location_sort", "footnotes_footnote"."doc_relation", "footnotes_footnote"."notes", "footnotes_footnote"."content", "footnotes_footnote"."url", "footnotes_footnote"."content_type_id", "footnotes_footnote"."object_id" FROM "footnotes_footnote" INNER JOIN "footnotes_source" ON ("footnotes_footnote"."source_id" = "footnotes_source"."id") WHERE ("footnotes_footnote"."content_type_id" = 8 AND "footnotes_footnote"."object_id" = 38281 AND "footnotes_footnote"."doc_relation"::text LIKE '%Y%') ORDER BY "footnotes_source"."title" ASC, "footnotes_source"."year" ASC LIMIT 21; args=(8, 38281, '%Y%')
[08/Aug/2023 15:19:35] DEBUG:django.db.backends::(0.000) SELECT "corpus_document"."doc_date_original", "corpus_document"."doc_date_calendar", "corpus_document"."doc_date_standard", "corpus_document"."id", "corpus_document"."image_order_override", "corpus_document"."shelfmark_override", "corpus_document"."description", "corpus_document"."description_en", "corpus_document"."description_he", "corpus_document"."description_ar", "corpus_document"."doctype_id", "corpus_document"."language_note", "corpus_document"."notes", "corpus_document"."created", "corpus_document"."last_modified", "corpus_document"."needs_review", "corpus_document"."old_pgpids", "corpus_document"."status" FROM "corpus_document" WHERE "corpus_document"."id" = 38281 LIMIT 21; args=(38281,)
[08/Aug/2023 15:19:35] DEBUG:django.db.backends::(0.001) SELECT "corpus_textblock"."id", "corpus_textblock"."document_id", "corpus_textblock"."fragment_id", "corpus_textblock"."certain", "corpus_textblock"."selected_images", "corpus_textblock"."region", "corpus_textblock"."multifragment", "corpus_textblock"."order" FROM "corpus_textblock" WHERE "corpus_textblock"."document_id" = 38281 ORDER BY "corpus_textblock"."order" ASC; args=(38281,)
[08/Aug/2023 15:19:35] DEBUG:django.db.backends::(0.001) SELECT "corpus_fragment"."id", "corpus_fragment"."shelfmark", "corpus_fragment"."old_shelfmarks", "corpus_fragment"."collection_id", "corpus_fragment"."url", "corpus_fragment"."iiif_url", "corpus_fragment"."is_multifragment", "corpus_fragment"."notes", "corpus_fragment"."needs_review", "corpus_fragment"."manifest_id", "corpus_fragment"."created", "corpus_fragment"."last_modified" FROM "corpus_fragment" WHERE "corpus_fragment"."id" = 13396 LIMIT 21; args=(13396,)

>>> d.default_translation
[08/Aug/2023 15:19:35] DEBUG:django.db.backends::(0.004) SELECT "footnotes_footnote"."id", "footnotes_footnote"."source_id", "footnotes_footnote"."location", "footnotes_footnote"."location_sort", "footnotes_footnote"."doc_relation", "footnotes_footnote"."notes", "footnotes_footnote"."content", "footnotes_footnote"."url", "footnotes_footnote"."content_type_id", "footnotes_footnote"."object_id" FROM "footnotes_footnote" INNER JOIN "footnotes_source" ON ("footnotes_footnote"."source_id" = "footnotes_source"."id") INNER JOIN "footnotes_source_languages" ON ("footnotes_source"."id" = "footnotes_source_languages"."source_id") INNER JOIN "footnotes_sourcelanguage" ON ("footnotes_source_languages"."sourcelanguage_id" = "footnotes_sourcelanguage"."id") WHERE ("footnotes_footnote"."content_type_id" = 8 AND "footnotes_footnote"."object_id" = 38281 AND "footnotes_footnote"."doc_relation"::text LIKE '%Y%' AND "footnotes_sourcelanguage"."code" = 'en') ORDER BY "footnotes_source"."title" ASC, "footnotes_source"."year" ASC LIMIT 1; args=(8, 38281, '%Y%', 'en')
[08/Aug/2023 15:19:35] DEBUG:django.db.backends::(0.000) SELECT "corpus_document"."doc_date_original", "corpus_document"."doc_date_calendar", "corpus_document"."doc_date_standard", "corpus_document"."id", "corpus_document"."image_order_override", "corpus_document"."shelfmark_override", "corpus_document"."description", "corpus_document"."description_en", "corpus_document"."description_he", "corpus_document"."description_ar", "corpus_document"."doctype_id", "corpus_document"."language_note", "corpus_document"."notes", "corpus_document"."created", "corpus_document"."last_modified", "corpus_document"."needs_review", "corpus_document"."old_pgpids", "corpus_document"."status" FROM "corpus_document" WHERE "corpus_document"."id" = 38281 LIMIT 21; args=(38281,)
[08/Aug/2023 15:19:35] DEBUG:django.db.backends::(0.000) SELECT "corpus_textblock"."id", "corpus_textblock"."document_id", "corpus_textblock"."fragment_id", "corpus_textblock"."certain", "corpus_textblock"."selected_images", "corpus_textblock"."region", "corpus_textblock"."multifragment", "corpus_textblock"."order" FROM "corpus_textblock" WHERE "corpus_textblock"."document_id" = 38281 ORDER BY "corpus_textblock"."order" ASC; args=(38281,)
[08/Aug/2023 15:19:35] DEBUG:django.db.backends::(0.000) SELECT "corpus_fragment"."id", "corpus_fragment"."shelfmark", "corpus_fragment"."old_shelfmarks", "corpus_fragment"."collection_id", "corpus_fragment"."url", "corpus_fragment"."iiif_url", "corpus_fragment"."is_multifragment", "corpus_fragment"."notes", "corpus_fragment"."needs_review", "corpus_fragment"."manifest_id", "corpus_fragment"."created", "corpus_fragment"."last_modified" FROM "corpus_fragment" WHERE "corpus_fragment"."id" = 13396 LIMIT 21; args=(13396,)
```

with prefecthing:

```sql
>>> d.digital_translations()
[08/Aug/2023 15:15:43] DEBUG:django.db.backends::(0.001) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."app_label" = 'corpus' AND "django_content_type"."model" = 'document') LIMIT 21; args=('corpus', 'document')
[08/Aug/2023 15:15:43] DEBUG:django.db.backends::(0.005) SELECT "footnotes_footnote"."id", "footnotes_footnote"."source_id", "footnotes_footnote"."location", "footnotes_footnote"."location_sort", "footnotes_footnote"."doc_relation", "footnotes_footnote"."notes", "footnotes_footnote"."content", "footnotes_footnote"."url", "footnotes_footnote"."content_type_id", "footnotes_footnote"."object_id", "footnotes_source"."id", "footnotes_source"."title", "footnotes_source"."title_en", "footnotes_source"."title_he", "footnotes_source"."title_ar", "footnotes_source"."year", "footnotes_source"."edition", "footnotes_source"."volume", "footnotes_source"."issue", "footnotes_source"."journal", "footnotes_source"."page_range", "footnotes_source"."publisher", "footnotes_source"."place_published", "footnotes_source"."other_info", "footnotes_source"."source_type_id", "footnotes_source"."url", "footnotes_source"."notes" FROM "footnotes_footnote" INNER JOIN "footnotes_source" ON ("footnotes_footnote"."source_id" = "footnotes_source"."id") WHERE ("footnotes_footnote"."content_type_id" = 8 AND "footnotes_footnote"."object_id" = 38281 AND "footnotes_footnote"."doc_relation"::text LIKE '%Y%') ORDER BY "footnotes_source"."title" ASC, "footnotes_source"."year" ASC LIMIT 21; args=(8, 38281, '%Y%')
[08/Aug/2023 15:15:43] DEBUG:django.db.backends::(0.001) SELECT ("footnotes_source_languages"."source_id") AS "_prefetch_related_val_source_id", "footnotes_sourcelanguage"."id", "footnotes_sourcelanguage"."name", "footnotes_sourcelanguage"."code", "footnotes_sourcelanguage"."direction" FROM "footnotes_sourcelanguage" INNER JOIN "footnotes_source_languages" ON ("footnotes_sourcelanguage"."id" = "footnotes_source_languages"."sourcelanguage_id") WHERE "footnotes_source_languages"."source_id" IN (29); args=(29,)
[08/Aug/2023 15:15:43] DEBUG:django.db.backends::(0.000) SELECT "corpus_document"."doc_date_original", "corpus_document"."doc_date_calendar", "corpus_document"."doc_date_standard", "corpus_document"."id", "corpus_document"."image_order_override", "corpus_document"."shelfmark_override", "corpus_document"."description", "corpus_document"."description_en", "corpus_document"."description_he", "corpus_document"."description_ar", "corpus_document"."doctype_id", "corpus_document"."language_note", "corpus_document"."notes", "corpus_document"."created", "corpus_document"."last_modified", "corpus_document"."needs_review", "corpus_document"."old_pgpids", "corpus_document"."status" FROM "corpus_document" WHERE "corpus_document"."id" = 38281 LIMIT 21; args=(38281,)
[08/Aug/2023 15:15:43] DEBUG:django.db.backends::(0.000) SELECT "corpus_textblock"."id", "corpus_textblock"."document_id", "corpus_textblock"."fragment_id", "corpus_textblock"."certain", "corpus_textblock"."selected_images", "corpus_textblock"."region", "corpus_textblock"."multifragment", "corpus_textblock"."order" FROM "corpus_textblock" WHERE "corpus_textblock"."document_id" = 38281 ORDER BY "corpus_textblock"."order" ASC; args=(38281,)
[08/Aug/2023 15:15:43] DEBUG:django.db.backends::(0.000) SELECT "corpus_fragment"."id", "corpus_fragment"."shelfmark", "corpus_fragment"."old_shelfmarks", "corpus_fragment"."collection_id", "corpus_fragment"."url", "corpus_fragment"."iiif_url", "corpus_fragment"."is_multifragment", "corpus_fragment"."notes", "corpus_fragment"."needs_review", "corpus_fragment"."manifest_id", "corpus_fragment"."created", "corpus_fragment"."last_modified" FROM "corpus_fragment" WHERE "corpus_fragment"."id" = 13396 LIMIT 21; args=(13396,)

>>> d.default_translation
[08/Aug/2023 15:15:58] DEBUG:django.db.backends::(0.004) SELECT "footnotes_footnote"."id", "footnotes_footnote"."source_id", "footnotes_footnote"."location", "footnotes_footnote"."location_sort", "footnotes_footnote"."doc_relation", "footnotes_footnote"."notes", "footnotes_footnote"."content", "footnotes_footnote"."url", "footnotes_footnote"."content_type_id", "footnotes_footnote"."object_id", "footnotes_source"."id", "footnotes_source"."title", "footnotes_source"."title_en", "footnotes_source"."title_he", "footnotes_source"."title_ar", "footnotes_source"."year", "footnotes_source"."edition", "footnotes_source"."volume", "footnotes_source"."issue", "footnotes_source"."journal", "footnotes_source"."page_range", "footnotes_source"."publisher", "footnotes_source"."place_published", "footnotes_source"."other_info", "footnotes_source"."source_type_id", "footnotes_source"."url", "footnotes_source"."notes" FROM "footnotes_footnote" INNER JOIN "footnotes_source" ON ("footnotes_footnote"."source_id" = "footnotes_source"."id") INNER JOIN "footnotes_source_languages" ON ("footnotes_source"."id" = "footnotes_source_languages"."source_id") INNER JOIN "footnotes_sourcelanguage" ON ("footnotes_source_languages"."sourcelanguage_id" = "footnotes_sourcelanguage"."id") WHERE ("footnotes_footnote"."content_type_id" = 8 AND "footnotes_footnote"."object_id" = 38281 AND "footnotes_footnote"."doc_relation"::text LIKE '%Y%' AND "footnotes_sourcelanguage"."code" = 'en') ORDER BY "footnotes_source"."title" ASC, "footnotes_source"."year" ASC LIMIT 1; args=(8, 38281, '%Y%', 'en')
[08/Aug/2023 15:15:58] DEBUG:django.db.backends::(0.000) SELECT ("footnotes_source_languages"."source_id") AS "_prefetch_related_val_source_id", "footnotes_sourcelanguage"."id", "footnotes_sourcelanguage"."name", "footnotes_sourcelanguage"."code", "footnotes_sourcelanguage"."direction" FROM "footnotes_sourcelanguage" INNER JOIN "footnotes_source_languages" ON ("footnotes_sourcelanguage"."id" = "footnotes_source_languages"."sourcelanguage_id") WHERE "footnotes_source_languages"."source_id" IN (29); args=(29,)
[08/Aug/2023 15:15:58] DEBUG:django.db.backends::(0.000) SELECT "corpus_document"."doc_date_original", "corpus_document"."doc_date_calendar", "corpus_document"."doc_date_standard", "corpus_document"."id", "corpus_document"."image_order_override", "corpus_document"."shelfmark_override", "corpus_document"."description", "corpus_document"."description_en", "corpus_document"."description_he", "corpus_document"."description_ar", "corpus_document"."doctype_id", "corpus_document"."language_note", "corpus_document"."notes", "corpus_document"."created", "corpus_document"."last_modified", "corpus_document"."needs_review", "corpus_document"."old_pgpids", "corpus_document"."status" FROM "corpus_document" WHERE "corpus_document"."id" = 38281 LIMIT 21; args=(38281,)
[08/Aug/2023 15:15:58] DEBUG:django.db.backends::(0.000) SELECT "corpus_textblock"."id", "corpus_textblock"."document_id", "corpus_textblock"."fragment_id", "corpus_textblock"."certain", "corpus_textblock"."selected_images", "corpus_textblock"."region", "corpus_textblock"."multifragment", "corpus_textblock"."order" FROM "corpus_textblock" WHERE "corpus_textblock"."document_id" = 38281 ORDER BY "corpus_textblock"."order" ASC; args=(38281,)
[08/Aug/2023 15:15:58] DEBUG:django.db.backends::(0.000) SELECT "corpus_fragment"."id", "corpus_fragment"."shelfmark", "corpus_fragment"."old_shelfmarks", "corpus_fragment"."collection_id", "corpus_fragment"."url", "corpus_fragment"."iiif_url", "corpus_fragment"."is_multifragment", "corpus_fragment"."notes", "corpus_fragment"."needs_review", "corpus_fragment"."manifest_id", "corpus_fragment"."created", "corpus_fragment"."last_modified" FROM "corpus_fragment" WHERE "corpus_fragment"."id" = 13396 LIMIT 21; args=(13396,)
```

Reply:
In any case, I think we're talking about differences of one query here, so I'm going to go ahead and merge! I can open a new issue to improve prefetching later if needed.